src/V3EmitMk.cpp

```diff
@@ -31,11 +31,557 @@ class EmitMk final {
+        }
+        UINFO(LOG_LEVEL, std::setw(maxScoreWidth) << (topScore + 1) << endl);
+    }
+
+    // Debug logging: prints Work Lists and their lists of files
+    static void debugLogWorkLists(const std::vector<WorkList>& workLists) {
+        constexpr int LOG_LEVEL = 5;
+
+        if (debug() < LOG_LEVEL) return;
+
+        UINFO(5, "Initial Work Lists with their concatenation eligibility status:" << endl);
+        for (const auto& list : workLists) {
+            UINFO(5, logMsgPrefixListEntryWithStatus(list.isConcatenable, 0)
+                         << "Work List #" << list.dbgId << "  ("
+                         << "num of files: " << list.files.size() << "; "
+                         << "total score: " << list.totalScore << ")" << endl);
+            if (debug() >= 6) {
+                // Log all files
+                for (const auto& file : list.files)
+                    UINFO(6, logMsgPrefixListEntry(1, 1)
+                                 << file.name << "  ("
+                                 << "score: " << file.score << ")" << endl);
+            } else {
+                // Log only first and last file
+                if (list.files.size() > 3) {
+                    const auto& first = list.files.front();
+                    const auto& last = list.files.back();
+                    UINFO(5, logMsgPrefixListEntry(1, 1)
+                                 << first.name << "  ("
+                                 << "score: " << first.score << ")" << endl);
+                    UINFO(5, logMsgPrefixListEntry(1, 1)
+                                 << "(... " << (list.files.size() - 2) << " files ...)" << endl);
+                    UINFO(5, logMsgPrefixListEntry(1, 1)
+                                 << last.name << "  ("
+                                 << "score: " << last.score << ")" << endl);
+                } else {
+                    // Print full list, as it is as long or shorter than abbreviated list.
+                    for (const auto& file : list.files)
+                        UINFO(5, logMsgPrefixListEntry(1)
+                                     << file.name << "  ("
+                                     << "score: " << file.score << ")" << endl);
+                }
+            }
+        }
+    }
+
+    // Debug logging: prints list of output files. List of grouped files is additionally printed
+    // for each concatenating file.
+    static void
+    debugLogOutputFilesList(const std::vector<FileOrConcatenatedFilesList>& outputFiles,
+                            bool logToStdOut = false) {
+        constexpr int LOG_LEVEL = 5;
+
+        if (!logToStdOut && debug() < LOG_LEVEL) return;
+
+        std::stringstream line;
+
+// Prints to stdout or debug log, depending on value of `logToStdOut`.
+#define DEBUG_LOG_OUTPUT_FILE_LIST_MSG(stmt) \
+    line << stmt; \
+    if (!logToStdOut) \
+        UINFO(LOG_LEVEL, line.str()); \
+    else \
```

Comment:
This goes away with using a file, then I think you'll be able to remove the macro (inline what it does).

Reply:
Removed, as the test was a bit unstable.